### PR TITLE
Context Parallel Squashed MR for recipes dir

### PR DESCRIPTION
**Pull request by jomitchellnv**
Add context based parallelism to ESM2 through the addition of a CPAware Dataloader

- Adds a unit test: `test_cp_dataloader` file that tests the dataloader outputs.
- Adds CPAwareDataloader to `dataset.py` which brings in all the CP stuff
- Adds a training script for context parallel + DDP called `train_ddp_cp.py` that runs both DDP and CP.
- Adds updates to the collator file (docs)
- Adds CP docs to the readme.md (but will make a new .md file for CP soon)

Usage
`train_dataloader, dataset_or_sampler = create_cp_dataloader(dist_config, cp_world_size=torch.distributed.get_world_size(group=cp_group), cp_group=cp_group, cp_rank=cp_rank, **args.dataset)`

#### Usage

```python
TODO: Add code snippet
```

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [X] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests for bionemo2
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Pre-submit Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [ ] I have added/updated tests as needed
- [ ] All existing tests pass successfully

### CODE REVIEW

File: bionemo-recipes/recipes/esm2_native_te/README.md [hunk 1]
```diff
@@ -127,6 +127,27 @@ python train_fsdp2.py --config-name L0_sanity \
   use_sequence_packing=true
 ```
 
+### Context Parallelism
+
+We provide a training script [train_ddp_cp](./esm2_native_te/train_ddp_cp.py) and a sample config [L0_sanity_cp](./hydra_config/L0_sanity_cp.yaml) that uses context parallelism.
+
+In the config the argument `--cp_size` allows the user to set the size of the context parallel distributed group. When paired with Distributed Data Parallelism (DDP), the number of context parallel groups will be determined by `world_size//cp_size`.
+
+Thus, for example, if a user has 8 processes and sets `cp_size=2` they will have `2` CP groups and `4` DDP groups. During dataloading we make no assumptions about the data pipeline being deterministic or not. We simply unique data only for the DDP groups and select the relevant CP shards for the respective CP group.
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/README.md]: > We simply unique data

?

File: bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    train_dataloader, dataset_or_sampler = (
+        create_cp_dataloader(
+            dist_config,
+            cp_world_size=torch.distributed.get_world_size(group=cp_group),
+            cp_group=cp_group,
+            cp_rank=cp_rank,
+            **args.dataset,
+        )
+        if args.use_sequence_packing
+        else create_bshd_dataloader(dist_config, **args.dataset)
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py]: maybe just `assert args.use_sequence_packing, "CP training requires THD input format / sequence packing"`

File: bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        model = AutoModelForMaskedLM.from_config(config, trust_remote_code=True)
+
+    logger.info("Initialized Model:\n%s", model)
+
+    # The huggingface model has a contact head that we don't use in masked language pre-training, so we delete it to
+    # avoid errors with unused parameters.
+    try:
+        del model.esm.contact_head
+    except AttributeError:
+        pass
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py]: the HF model wont work here anyways, right? just delete this bit

**Review by pstjohn — APPROVED** [review]

File: bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py (hunk 3 above)
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py]: ack -- deleted

File: bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py (hunk 2 above)
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py]: added

File: bionemo-recipes/recipes/esm2_native_te/README.md (hunk 1 above)
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/README.md]: fixed

File: bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py (hunk 2 above)
Comment (pstjohn) [reply bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py]: i'm saying instead of creating the bshd dataloader here

Comment (pstjohn) [reply bionemo-recipes/recipes/esm2_native_te/train_ddp_cp.py]: oops missed the changes, looks great
